ssl/statem/extensions.c

```diff
@@ -623,7 +625,12 @@ int tls_collect_extensions(SSL *s, PACKET *packet, unsigned int context,
                 && type != TLSEXT_TYPE_cookie
                 && type != TLSEXT_TYPE_renegotiate
                 && type != TLSEXT_TYPE_signed_certificate_timestamp
-                && (s->ext.extflags[idx] & SSL_EXT_FLAG_SENT) == 0) {
+                && (s->ext.extflags[idx] & SSL_EXT_FLAG_SENT) == 0
+#ifndef OPENSSL_NO_GOST
+                && !(context & SSL_EXT_TLS1_2_SERVER_HELLO
+                     && type == TLSEXT_TYPE_cryptopro_bug)
```

Comment:
I'm looking at the whole expression, and it currently doesn't really make sense...  the first check of `context` will preclude the second check to ever happen...  so it turns out that the parentheses you had originally were actually correct, bug the operator wasn't!  Basically, when you wrote what can be summarised as `((A && B && C && D) && (GOST && CONDITION))`, what you really wanted was `((A && B && C && D) || (GOST && CONDITION))`, was it not?  I.e. the `&&` on line 630 should really have been `||`.  In that case, the parentheses I complained about actually *do* make sense.

Reply:
You will have to work on the indentation for this. Fair warning

Reply:
The extension 65000 is a rudiment from very old times. It was a way to demonstrate using non-standard PRF and server sends it even in a case when a client did not send it in ClientHello. So the original code enforces error in case of irrelevant extension outside of ClientHello and TLS 1.3-related context. I had to add an extra check to work around the initial weirdness of the ext65000.

Sorry, I do not understand how should the indentation look here. I've updated it according to your previous recommendation and can also run the script from util.

Reply:
Well, the || instead of && does not work, I get an error SSL_AD_UNSUPPORTED_EXTENSION

If I understand correctly, to use || instead of &&, we need to inverse the condition of the original check.

Reply:
Oh, right. So this? 

``` C
        if (idx < OSSL_NELEM(ext_defs)
                && (    ((context & (SSL_EXT_CLIENT_HELLO
                                     | SSL_EXT_TLS1_3_CERTIFICATE_REQUEST
                                     | SSL_EXT_TLS1_3_NEW_SESSION_TICKET)) == 0
                         && type != TLSEXT_TYPE_cookie
                         && type != TLSEXT_TYPE_renegotiate
                         && type != TLSEXT_TYPE_signed_certificate_timestamp
                         && (s->ext.extflags[idx] & SSL_EXT_FLAG_SENT) == 0)
#ifndef OPENSSL_NO_GOST
                    || ((context & SSL_EXT_TLS1_2_SERVER_HELLO) == 0
                         && type != TLSEXT_TYPE_cryptopro_bug)
#endif
								)) {

Reply:
I strongly suspect that our discussion is a good indicator that this piece of code is not clear...

Reply:
Somehow, I can't make sense of the logic.  Basically, your added expression could as well be written like this:

``` C
                && ((context & SSL_EXT_TLS1_2_SERVER_HELLO) == 0
                     || type != TLSEXT_TYPE_cryptopro_bug)
```

... and compared to what's there before your change, it looks...  odd.  I comment to explain why your logic deviates from the rest of the expression might be a good thing...

Reply:
So well, I've tried all the variants of fix you've suggested (and some more) and they do not work :(

Reply:
The logic as written looks correct to me. (except I would add `!= 0` onto the end of `(context & SSL_EXT_TLS1_2_SERVER_HELLO)` for conformance with our style).

The change in this part of the code seems to be a bug fix not related to the problem described in #7747, so perhaps it should be in a separate commit.

Reply:
Just wondering: Is this really the right location for the fix? The original report in #7747 was talking about an **illegal_parameter**, not an **unsupported_extension** alert.

Reply:
Hmm strange... I didn't see Matt's comment when I posted my last comment, which apparently explains why I was wondering.

Reply:
Oops. It seems to be a separate problem, so the original fix recommended by @mattcaswell has nothing to do with it. I removed it from the PR.

On commit to upstream the message should be changed.

Reply:
@bedmit is this really a fix for an unrelated error (as Matt suspects) or is it a misplaced fix (as I suspect)?

Reply:
@mspncp a fix to an unrelated error.

Reply:
The idea I mean is: we shouldn't throw SSLfatal when 
1) We are in ServerHello
2) We got CryptoPro extension.

Reply:
Arghh... the last `==` is an `!=`:

```
#ifndef OPENSSL_NO_GOST
                && (context & SSL_EXT_TLS1_2_SERVER_HELLO == 0
                     || type != TLSEXT_TYPE_cryptopro_bug)
#endif
```
(Time to stop for today; too much boolean algebra is hazadardous for the human brain, see https://github.com/openssl/openssl/pull/6880#issuecomment-410938813)

Reply:
> But alltogether, it still does not make sense to me. (In particular the last ||).

It makes sense to me :-)

This whole `if` is about throwing a fatal error if we got an unsolicited extension. Except that in some circumstances unsolicited extensions are ok, so we have to check all of these. The cryptopro_bug extension is one of those. It can appear in a TLSv1.2 ServerHello even though it didn't appear in the ClientHello.

So if what we've got isn't a TLSv1.2 ServerHello, or isn't the cryptopro_bug extension then we should throw an alert (or equivalently we shouldn't throw an alert if it *is* a TLSv1.2 ServerHello *and* it is the cryptopro_bug extension).

Reply:
Thank you @beldmit and @mattcaswell for your patient explanations. Now it makes sense to me, too.

Reply:
@levitte, is the change request still valid?

Reply:
No, this is good. I still haven't wrapped my brain around it, and won't today.
I trust @mattcasswell here, as this is his domain much more than mine, and in particular in view of his explanation, there's nothing more for me to say.